gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java

```diff
@@ -126,58 +139,88 @@ static GoogleCloudStorageGrpcReadChannel open(
 
   private static GoogleCloudStorageGrpcReadChannel openChannel(
       StorageStubProvider stubProvider,
+      Storage storage,
+      ApiErrorExtractor errorExtractor,
       StorageResourceId resourceId,
       GoogleCloudStorageReadOptions readOptions,
       BackOffFactory backOffFactory) throws IOException {
     StorageBlockingStub stub = stubProvider.newBlockingStub();
     // TODO(b/135138893): We can avoid this call by adding metadata to a read request.
     //      That will save about 40ms per read.
-    com.google.google.storage.v1.Object storageObject = getObjectMetadata(resourceId, readOptions,
-        stub);
+    Preconditions.checkArgument(storage != null, "GCS json client cannot be null");
+    GoogleCloudStorageItemInfo itemInfo = getObjectMetadata(resourceId,
+        errorExtractor, backOffFactory, storage);
+    Preconditions.checkArgument(itemInfo != null, "object metadata cannot be null");
+    // The non-gRPC read channel has special support for gzip. This channel doesn't
+    // decompress gzip-encoded objects on the fly, so best to fail fast rather than return
+    // gibberish unexpectedly.
+    String contentEncoding = itemInfo.getContentEncoding();
+    if (contentEncoding != null && contentEncoding.contains("gzip")) {
+      throw new IOException(
+          "Cannot read GZIP encoded files - content encoding support is disabled.");
+    }
 
-    Preconditions.checkArgument(storageObject != null, "object metadata cannot be null");
     int prefetchSizeInBytes = readOptions.getMinRangeRequestSize() / 2;
-    long footerOffsetInBytes = Math.max(0, (storageObject.getSize() - prefetchSizeInBytes));
+    long footerOffsetInBytes = Math.max(0, (itemInfo.getSize() - prefetchSizeInBytes));
 
     ByteString footerContent = getFooterContent(resourceId, readOptions, stub, footerOffsetInBytes);
 
     return new GoogleCloudStorageGrpcReadChannel(
         stub,
         stubProvider,
         resourceId,
-        storageObject.getGeneration(),
-        storageObject.getSize(),
+        itemInfo.getContentGeneration(),
+        itemInfo.getSize(),
         footerOffsetInBytes,
         footerContent,
         readOptions,
         backOffFactory);
   }
 
-  private static com.google.google.storage.v1.Object getObjectMetadata(StorageResourceId resourceId,
-      GoogleCloudStorageReadOptions readOptions, StorageBlockingStub stub) throws IOException {
+  private static GoogleCloudStorageItemInfo getObjectMetadata(
+      StorageResourceId resourceId,
+      ApiErrorExtractor errorExtractor,
+      BackOffFactory backOffFactory,
+      Storage gcs) throws IOException {
+    StorageObject object;
     try {
-      // TODO(b/151184800): Implement per-message timeout, in addition to stream timeout.
```

Comment:
Once read meta data is supported on gRPC, we are going to revert back to this logic right? So should we delete this piece of code or just hide it behind a hard coded flag?

Reply:
I prefer reverting the commit to flags, since we don't know the exact time lines of the switch back to gRPC for metadata APIs